tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_graphics_GC.java

```diff
@@ -246,6 +248,60 @@ public void test_dispose() {
 	gc.dispose();
 }
 
+
+@Test
+public void test_drawImage_nonAutoScalableGC_bug_2504() {
+	Shell shell = new Shell(display);
+    shell.setLayout(new FormLayout());
+
+    float targetScale = 2f;
+    int srcSize = 50;
+
+    Image image = new Image(display, srcSize, srcSize);
+    GC gcSrc = new GC(image);
+    gcSrc.setBackground(display.getSystemColor(SWT.COLOR_BLACK));
+    gcSrc.fillRectangle(0, 0, srcSize, srcSize);
+    gcSrc.setBackground(display.getSystemColor(SWT.COLOR_WHITE));
+    gcSrc.fillRectangle(2, 2, srcSize - 4, srcSize - 4);
+    gcSrc.dispose();
+
+    Rectangle bounds = image.getBounds();
+
+    Canvas canvas = new Canvas(shell, SWT.NONE) {
+        @Override
+        public boolean isAutoScalable() {
+            return false;
+        }
+    };
+    FormData formData = new FormData();
+    formData.width = Math.round(bounds.width * targetScale);
+    formData.height = Math.round(bounds.height * targetScale);
+    canvas.setLayoutData(formData);
+    canvas.addPaintListener(e -> {
+        e.gc.drawImage(image, 0, 0, bounds.width, bounds.height,
+                       0, 0, formData.width, formData.height);
+    });
+
+    shell.open();
+    while (display.readAndDispatch()) {
+    }
```

Comment:
Why is this necessary? Can't we remove that?

Reply:
@arunjose696 Can you explain where this risk come from? `shell.open()` will send a paint event and trigger the paint listener of the canvas, so I would expect the `copyArea()` operation to work. What will the `readAndDispatch()` do that leads to "fully" painting the canvas? Is there any way to produce the issue you describe?